api/cluster.go

```diff
@@ -25,5 +29,82 @@ func (s *Server) appStatus(ctx *middleware.Context) {
 		return
 	}
 
-	response.Write(ctx, response.NewError(http.StatusServiceUnavailable, errors.New("node not ready")))
+	response.Write(ctx, response.NewError(http.StatusServiceUnavailable, "node not ready"))
+}
+
+func (s *Server) getClusterStatus(ctx *middleware.Context) {
+	status := models.ClusterStatus{
+		Node:  cluster.ThisNode,
+		Peers: cluster.GetPeers(),
+	}
+	response.Write(ctx, response.NewJson(200, status, ""))
+}
+
+// IndexFind returns a sequence of msgp encoded idx.Node's
+func (s *Server) indexFind(ctx *middleware.Context, req models.IndexFind) {
+	// metricDefs only get updated periodically (when using CassandraIdx), so we add a 1day (86400seconds) buffer when
+	// filtering by our From timestamp.  This should be moved to a configuration option
+	if req.From != 0 {
+		req.From -= 86400
+	}
+	resp := models.NewIndexFindResp()
+
+	for _, pattern := range req.Patterns {
+		nodes, err := s.MetricIndex.Find(req.OrgId, pattern, req.From)
+		if err != nil {
+			response.Write(ctx, response.NewError(http.StatusBadRequest, err.Error()))
+			return
+		}
+		resp.Nodes[pattern] = nodes
+	}
+	response.Write(ctx, response.NewMsgp(200, resp))
+}
+
+// IndexGet returns a msgp encoded schema.MetricDefinition
+func (s *Server) indexGet(ctx *middleware.Context, req models.IndexGet) {
+	def, err := s.MetricIndex.Get(req.Id)
+	if err != nil {
+	} else { // currently this can only be notFound
+		response.Write(ctx, response.NewError(http.StatusNotFound, "Not Found"))
+		return
+	}
+
+	response.Write(ctx, response.NewMsgp(200, &def))
+}
+
+// IndexList returns msgp encoded schema.MetricDefinition's
+func (s *Server) indexList(ctx *middleware.Context, req models.IndexList) {
+	defs := s.MetricIndex.List(req.OrgId)
+	resp := make([]msgp.Marshaler, len(defs))
+	for i, _ := range defs {
+		d := defs[i]
+		resp[i] = &d
+	}
+	response.Write(ctx, response.NewMsgpArray(200, resp))
+}
+
+func (s *Server) getData(ctx *middleware.Context, request models.GetData) {
+	series, err := s.getTargetsLocal(request.Requests)
+	if err != nil {
+		// the only errors returned are from us catching panics, so we should treat them
+		// all as internalServerErrors
+		log.Error(3, "HTTP getData() %s", err.Error())
+		response.Write(ctx, response.WrapError(err))
+		return
+	}
+	response.Write(ctx, response.NewMsgp(200, &models.GetDataResp{Series: series}))
+}
+
+func (s *Server) indexDelete(ctx *middleware.Context, req models.IndexDelete) {
+	defs, err := s.MetricIndex.Delete(req.OrgId, req.Query)
+	if err != nil {
+		// errors can only be caused by bad request.
```

Comment:
another case of callee implementation leaking into caller, where I think more specific, detailed errors should flow from the code that actually generates them.